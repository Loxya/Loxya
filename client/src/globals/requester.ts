import Requester from '@/utils/requester';

const requester = new Requester();

export const init = (): void => {
    requester.init();
};

export * from '@/utils/requester';
export default requester;
