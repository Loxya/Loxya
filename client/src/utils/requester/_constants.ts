export { HttpStatusCode as HttpCode } from 'axios';

//
// - Built-in error codes
//

export enum ClientErrorCode {
    NETWORK_ERROR = 'CLIENT_NETWORK_ERROR',
    TIMEOUT_ERROR = 'CLIENT_TIMEOUT_ERROR',
}
