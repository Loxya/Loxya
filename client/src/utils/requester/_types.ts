import type { ClientErrorCode } from './_constants';
import type { ApiErrorCode } from '@/stores/api/@codes';
import type {
    AxiosResponse,
    AxiosRequestConfig,
    AxiosResponseHeaders,
    RawAxiosResponseHeaders,
} from 'axios';

export type ResponseHeaders = RawAxiosResponseHeaders | AxiosResponseHeaders;

//
// - Request
//

export type ProgressCallback = (percent: number) => void;

export type RequestConfig<D = any> = (
    & AxiosRequestConfig<D>
    & { onProgress?: ProgressCallback }
);

export type RequestXhrPromise<Data = never> = Promise<AxiosResponse<Data>>;

//
// - Errors
//

export type ErrorCode = ClientErrorCode | ApiErrorCode;
