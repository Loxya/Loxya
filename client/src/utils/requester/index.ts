export type {
    ErrorCode,
    RequestConfig,
    ResponseHeaders,
    ProgressCallback,
} from './_types';

export { RequestError } from './_helpers';
export { HttpCode, ClientErrorCode } from './_constants';
export { default } from './requester';
