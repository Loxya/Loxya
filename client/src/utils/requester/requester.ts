import axios from 'axios';
import invariant from 'invariant';
import globalConfig from '@/globals/config';
import interceptors from './_interceptors';

import type { AxiosInstance, AxiosResponse } from 'axios';
import type { RequestConfig, RequestXhrPromise, ResponseHeaders } from './_types';

class Requester {
    /** Instance d'Axios. */
    private _instance: AxiosInstance | undefined;

    /** L'initialisation a-t-elle déjà été faite ? */
    private _isInitialized: boolean;

    /**
     * Retourne l'état d'initialisation de `Requester`.
     *
     * @returns L'état d'initialisation de `Requester`.
     */
    public get isInitialized(): boolean {
        return this._isInitialized;
    }

    /**
     * Point d'entrée vers la propriété `interceptors` d'Axios.
     *
     * @returns L'objet contenant les intercepteurs de requête/réponse d'Axios.
     */
    public get interceptors(): AxiosInstance['interceptors'] {
        invariant(
            (this._isInitialized && this._instance),
            '[Requester] Initialization (`Requester.init({ ... })`) must be ' +
            'done prior to any modifications to the interceptors.',
        );
        return this._instance.interceptors;
    }

    /**
     * Constructeur.
     *
     * @param config - Si défini, déclenche l'initialisation immédiate (`Requester.init()`)
     *                 avec la configuration passée.
     */
    constructor(config?: RequestConfig) {
        this._isInitialized = false;

        if (config) {
            this.init(config);
        }
    }

    /**
     * Initialisation.
     *
     * @param config - L'objet littéral contenant la configuration de base
     *                 utilisée pour toutes les requêtes.
     *                 correspond à la configuration supportée par Axios.
     */
    public init(config: RequestConfig = {}): void {
        invariant(!this._isInitialized, '[Requester] Unexpected duplicate initialization.');
        this._isInitialized = true;

        // - Création de l'instance.
        const defaultConfig: RequestConfig = {
            baseURL: globalConfig.api.url,
            headers: globalConfig.api.headers,
        };
        this._instance = axios.create({ ...defaultConfig, ...config });

        // - Intercepteurs
        (['request', 'response'] as const).forEach((type) => {
            if (!(type in interceptors)) {
                return;
            }

            const interceptor = !Array.isArray(interceptors[type])
                ? [interceptors[type], (err: unknown) => Promise.reject(err)]
                : interceptors[type];

            this.interceptors[type].use(...interceptor as any[]);
        });
    }

    /**
     * Exécute une requête HTTP.
     *
     * @param config - La configuration de la requête contenant l'URL, la méthode à utiliser, les données éventuelles, etc.
     *                 Les valeurs par défaut sont celles passées lors de l'initialisation.
     *
     * @returns Une promesse résolue avec un objet contenant les informations de réponse
     *          sous la forme : `{ headers: {...}, data: {...}, ... }`.
     */
    public exec<T = never>(config: RequestConfig): RequestXhrPromise<T> {
        invariant(
            this._isInitialized,
            '[Requester] The initialization (`Request.init({ ... })`) must be ' +
            'done before calling the query methods.',
        );
        return this._instance!.request<T>(config);
    }
}

//
// - Méthodes HTTP utilitaires
//

interface Requester {
    head(url: string, config?: RequestConfig): Promise<ResponseHeaders>;
    options(url: string, config?: RequestConfig): Promise<ResponseHeaders>;

    get<Data = unknown>(url: string, config?: RequestConfig): Promise<Data>;
    delete<Data = unknown>(url: string, config?: RequestConfig): Promise<Data>;

    post<Data = unknown>(url: string, data?: unknown, config?: RequestConfig): Promise<Data>;
    put<Data = unknown>(url: string, data?: unknown, config?: RequestConfig): Promise<Data>;
    patch<Data = unknown>(url: string, data?: unknown, config?: RequestConfig): Promise<Data>;

    postForm<Data = unknown>(url: string, data?: unknown, config?: RequestConfig): Promise<Data>;
    putForm<Data = unknown>(url: string, data?: unknown, config?: RequestConfig): Promise<Data>;
    patchForm<Data = unknown>(url: string, data?: unknown, config?: RequestConfig): Promise<Data>;
}

(['head', 'options'] as const).forEach((method) => {
    // eslint-disable-next-line func-names
    Requester.prototype[method] = function(
        url: string,
        config: RequestConfig = {},
    ): Promise<ResponseHeaders> {
        const xhrPromise = this.exec({ ...config, method, url });
        return xhrPromise.then(({ headers }: AxiosResponse): ResponseHeaders => headers);
    };
});

(['delete', 'get'] as const).forEach((method) => {
    // eslint-disable-next-line func-names
    Requester.prototype[method] = function<T = never>(
        url: string,
        config: RequestConfig = {},
    ): Promise<T> {
        const xhrPromise = this.exec<T>({ ...config, method, url });
        return xhrPromise.then(({ data }: AxiosResponse<T>): T => data);
    };
});

(['post', 'put', 'patch'] as const).forEach((baseMethod) => {
    ([baseMethod, `${baseMethod}Form`] as const).forEach((method) => {
        // eslint-disable-next-line func-names
        Requester.prototype[method] = function<T = never>(
            url: string,
            data: unknown = {},
            config: RequestConfig<T> = {},
        ): Promise<T> {
            const xhrPromise = this.exec<T>({ ...config, method, data, url });
            return xhrPromise.then(({ data: _data }: AxiosResponse<T>): T => _data);
        };
    });
});

export default Requester;
