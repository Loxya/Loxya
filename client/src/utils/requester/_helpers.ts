/* eslint-disable import/prefer-default-export */

import invariant from 'invariant';
import isValidInteger from '@/utils/isValidInteger';
import parseInteger from '@/utils/parseInteger';
import axios, { AxiosError } from 'axios';
import isPlainObject from 'lodash/isPlainObject';
import { ApiErrorCode } from '@/stores/api/@codes';
import { ClientErrorCode } from './_constants';

import type { AxiosResponse } from 'axios';
import type { HttpCode } from './_constants';
import type { ErrorCode } from './_types';

type ErrorPayload = {
    code: ErrorCode,
    message: string | undefined,
    details: unknown | undefined,
};

const normalizeErrorPayload = (payload: unknown): ErrorPayload => {
    const normalizedErrorPayload = isPlainObject(payload) ? payload : {};
    const { code: rawCode, message: rawMessage, ...extraFields } = normalizedErrorPayload;

    const code: ApiErrorCode = (() => {
        if (isValidInteger(rawCode)) {
            const normalizedCode = parseInteger(rawCode)!;
            if (Object.values(ApiErrorCode).includes(normalizedCode as any)) {
                return parseInteger(rawCode)! as ApiErrorCode;
            }
        }

        // eslint-disable-next-line no-console
        console.warn(
            'A response with an unknown `ApiErrorCode` ' +
            `has been returned: ${String(rawCode)}.`,
        );

        return ApiErrorCode.UNKNOWN;
    })();

    const message = typeof rawMessage === 'string'
        ? rawMessage
        : undefined;

    const details = Object.keys(extraFields).length > 0
        ? ('details' in extraFields ? extraFields.details : extraFields)
        : undefined;

    return { code, message, details };
};

export class RequestError extends Error {
    public readonly code: ErrorCode;

    public readonly httpCode: HttpCode | undefined;

    public readonly details: any | undefined;

    constructor(
        code: ErrorCode,
        message?: string,
        cause?: ErrorOptions['cause'],
        details?: any,
        httpCode?: HttpCode,
    ) {
        message ??= httpCode ? `Request failed with status code ${httpCode}` : 'Request failed';
        super(message, { cause });

        this.name = 'RequestError';
        this.code = code;
        this.details = details;
        this.httpCode = httpCode;
    }

    //
    // - Factories.
    //

    public static fromAxiosResponse({ status, data: payload }: AxiosResponse, cause?: ErrorOptions['cause']): RequestError {
        invariant(
            status < 200 || status > 299,
            'Error: Trying to create a request error from a successful response.',
        );

        const { code, message, details } = normalizeErrorPayload(payload?.error ?? payload);
        return new RequestError(code, message, cause, details, status as HttpCode);
    }

    public static fromAxiosError(error: AxiosError): RequestError {
        if (error.response) {
            return RequestError.fromAxiosResponse(error.response, error);
        }

        let code: ErrorCode = ApiErrorCode.UNKNOWN;
        if (AxiosError.ERR_NETWORK === error.code) {
            code = ClientErrorCode.NETWORK_ERROR;
        }
        if ([AxiosError.ECONNABORTED, AxiosError.ETIMEDOUT].includes(error.code as any)) {
            code = ClientErrorCode.TIMEOUT_ERROR;
        }
        return new RequestError(code, undefined, error);
    }

    public static fromError(error: unknown): RequestError {
        if (axios.isAxiosError(error)) {
            return RequestError.fromAxiosError(error);
        }
        return new RequestError(ApiErrorCode.UNKNOWN, undefined, error);
    }
}
