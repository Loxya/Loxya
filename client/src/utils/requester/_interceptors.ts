import axios from 'axios';
import deepObjectSet from 'lodash/set';
import { RequestError } from './_helpers';
import isPlainObject from 'lodash/isPlainObject';
import flattenObject from '@/utils/flattenObject';
import globalConfig from '@/globals/config';
import cookies from '@/utils/cookies';
import DateTime from '@/utils/datetime';
import Day from '@/utils/day';

import type { RequestConfig } from './_types';
import type {
    AxiosInterceptorManager,
    AxiosProgressEvent,
    AxiosResponse,
} from 'axios';

type Interceptor<T> = (
    | Parameters<AxiosInterceptorManager<T>['use']>
    | Parameters<AxiosInterceptorManager<T>['use']>[0]
);

type Interceptors = {
    request?: Interceptor<RequestConfig>,
    response?: Interceptor<AxiosResponse>,
};

const requestInterceptor: Interceptor<RequestConfig> = (rawConfig: RequestConfig): RequestConfig => {
    const { params, onProgress, ...others } = rawConfig;
    const config: RequestConfig = { ...others };

    // - Ajoute une méthode haut-niveau permettant de récupérer l'avancement sous forme numérique.
    //   - Si la méthode de la requête est POST, PUT ou PATCH, c'est l'événement `onUploadProgress` qui sera écouté.
    //   - Sinon ce sera `onDownloadProgress`
    if (onProgress) {
        const progressCallback = (event: AxiosProgressEvent): void => {
            if (!event.lengthComputable) {
                return;
            }

            const { loaded, total } = event;
            if (total === undefined) {
                return;
            }

            const percent = (loaded / total) * 100;
            onProgress(percent);
        };

        if (config.method && ['POST', 'PUT', 'PATCH'].includes(config.method.toUpperCase())) {
            config.onUploadProgress = progressCallback;
        } else {
            config.onDownloadProgress = progressCallback;
        }
    }

    // - Traitement spécial des objets littéraux contenant des fichiers.
    //   => On utilise un envoi via `multipart/form-data` et on met les données
    //      supplémentaires éventuelles dans une clé spéciale `@data`)
    const hasFiles = (
        isPlainObject(config.data) &&
        Object.values(config.data).some((value: unknown) => (
            value instanceof File
        ))
    );
    if (hasFiles) {
        const formData = new FormData();
        const data = Object.entries(flattenObject(config.data)).reduce(
            (_data: Record<string, unknown>, [key, value]: [string, unknown]) => {
                if (value instanceof File) {
                    formData.append(btoa(key), value);
                    return _data;
                }
                return deepObjectSet(_data, key, value);
            },
            {},
        );

        formData.append('@data', JSON.stringify(data));
        config.data = formData;
    }

    if (params) {
        Object.keys(params).forEach((name: keyof typeof params) => {
            if (params[name] === true) {
                params[name] = '1';
            }
            if (params[name] === false) {
                params[name] = '0';
            }
            if (params[name] instanceof DateTime || params[name] instanceof Day) {
                params[name] = params[name].toString();
            }
        });
        config.params = params;
    }

    const authToken = cookies.get(globalConfig.auth.cookie);
    if (authToken) {
        config.headers ??= {};
        config.headers.Authorization = `Bearer ${authToken}`;
    }

    return config;
};

const responseInterceptor: Interceptor<AxiosResponse> = [
    (response: AxiosResponse): any => response,
    (error: unknown): Promise<any> => {
        if (axios.isCancel(error)) {
            throw error;
        }
        return Promise.reject(RequestError.fromError(error));
    },
];

const interceptors: Interceptors = {
    request: requestInterceptor,
    response: responseInterceptor,
};

export default interceptors;
