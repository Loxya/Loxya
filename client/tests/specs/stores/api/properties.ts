import requester from '@/globals/requester';
import apiProperties from '@/stores/api/properties';
import data from '@fixtures/properties';

describe('Properties Api', () => {
    describe('all()', () => {
        it('parse the returned data correctly', async () => {
            jest.spyOn(requester, 'get').mockResolvedValue(data.details());
            await expect(apiProperties.all()).resolves.toMatchSnapshot();
        });
    });

    describe('one()', () => {
        it.each(data.details())('parse the returned data correctly (with #$id)', async (datum: any) => {
            jest.spyOn(requester, 'get').mockResolvedValue(datum);
            await expect(apiProperties.one(datum.id)).resolves.toMatchSnapshot();
        });
    });

    describe('create()', () => {
        it.each(data.details())('parse the returned data correctly (with #$id)', async (datum: any) => {
            jest.spyOn(requester, 'post').mockResolvedValue(datum);
            await expect(apiProperties.create({} as any)).resolves.toMatchSnapshot();
        });
    });

    describe('update()', () => {
        it.each(data.details())('parse the returned data correctly (with #$id)', async (datum: any) => {
            jest.spyOn(requester, 'put').mockResolvedValue(datum);
            await expect(apiProperties.update(datum.id, {} as any)).resolves.toMatchSnapshot();
        });
    });
});
