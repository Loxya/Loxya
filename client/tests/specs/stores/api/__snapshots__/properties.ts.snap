// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Properties Api all() parse the returned data correctly 1`] = `
[
  {
    "categories": [
      {
        "id": 2,
        "name": "Lumière",
      },
      {
        "id": 1,
        "name": "Son",
      },
    ],
    "entities": [
      "material",
    ],
    "id": 1,
    "is_totalisable": true,
    "name": "Poids",
    "type": "float",
    "unit": "kg",
  },
  {
    "categories": [],
    "entities": [
      "material",
    ],
    "id": 2,
    "max_length": null,
    "name": "Couleur",
    "type": "string",
  },
  {
    "categories": [
      {
        "id": 2,
        "name": "Lumière",
      },
      {
        "id": 1,
        "name": "Son",
      },
    ],
    "entities": [
      "material",
    ],
    "id": 3,
    "is_totalisable": true,
    "name": "Puissance",
    "type": "integer",
    "unit": "W",
  },
  {
    "categories": [],
    "entities": [
      "material",
    ],
    "id": 4,
    "name": "Conforme",
    "type": "boolean",
  },
  {
    "categories": [],
    "entities": [
      "material",
    ],
    "id": 5,
    "name": "Date d'achat",
    "type": "date",
  },
  {
    "categories": [],
    "entities": [
      "material",
    ],
    "id": 6,
    "is_totalisable": true,
    "name": "Prix d'achat",
    "type": "float",
    "unit": "€",
  },
  {
    "categories": [],
    "entities": [
      "material",
    ],
    "id": 7,
    "name": "Lieu d'usage",
    "options": [
      "Intérieur",
      "Extérieur",
      "Polyvalent",
    ],
    "type": "list",
  },
]
`;

exports[`Properties Api create() parse the returned data correctly (with #1) 1`] = `
{
  "categories": [
    {
      "id": 2,
      "name": "Lumière",
    },
    {
      "id": 1,
      "name": "Son",
    },
  ],
  "entities": [
    "material",
  ],
  "id": 1,
  "is_totalisable": true,
  "name": "Poids",
  "type": "float",
  "unit": "kg",
}
`;

exports[`Properties Api create() parse the returned data correctly (with #2) 1`] = `
{
  "categories": [],
  "entities": [
    "material",
  ],
  "id": 2,
  "max_length": null,
  "name": "Couleur",
  "type": "string",
}
`;

exports[`Properties Api create() parse the returned data correctly (with #3) 1`] = `
{
  "categories": [
    {
      "id": 2,
      "name": "Lumière",
    },
    {
      "id": 1,
      "name": "Son",
    },
  ],
  "entities": [
    "material",
  ],
  "id": 3,
  "is_totalisable": true,
  "name": "Puissance",
  "type": "integer",
  "unit": "W",
}
`;

exports[`Properties Api create() parse the returned data correctly (with #4) 1`] = `
{
  "categories": [],
  "entities": [
    "material",
  ],
  "id": 4,
  "name": "Conforme",
  "type": "boolean",
}
`;

exports[`Properties Api create() parse the returned data correctly (with #5) 1`] = `
{
  "categories": [],
  "entities": [
    "material",
  ],
  "id": 5,
  "name": "Date d'achat",
  "type": "date",
}
`;

exports[`Properties Api create() parse the returned data correctly (with #6) 1`] = `
{
  "categories": [],
  "entities": [
    "material",
  ],
  "id": 6,
  "is_totalisable": true,
  "name": "Prix d'achat",
  "type": "float",
  "unit": "€",
}
`;

exports[`Properties Api create() parse the returned data correctly (with #7) 1`] = `
{
  "categories": [],
  "entities": [
    "material",
  ],
  "id": 7,
  "name": "Lieu d'usage",
  "options": [
    "Intérieur",
    "Extérieur",
    "Polyvalent",
  ],
  "type": "list",
}
`;

exports[`Properties Api one() parse the returned data correctly (with #1) 1`] = `
{
  "categories": [
    {
      "id": 2,
      "name": "Lumière",
    },
    {
      "id": 1,
      "name": "Son",
    },
  ],
  "entities": [
    "material",
  ],
  "id": 1,
  "is_totalisable": true,
  "name": "Poids",
  "type": "float",
  "unit": "kg",
}
`;

exports[`Properties Api one() parse the returned data correctly (with #2) 1`] = `
{
  "categories": [],
  "entities": [
    "material",
  ],
  "id": 2,
  "max_length": null,
  "name": "Couleur",
  "type": "string",
}
`;

exports[`Properties Api one() parse the returned data correctly (with #3) 1`] = `
{
  "categories": [
    {
      "id": 2,
      "name": "Lumière",
    },
    {
      "id": 1,
      "name": "Son",
    },
  ],
  "entities": [
    "material",
  ],
  "id": 3,
  "is_totalisable": true,
  "name": "Puissance",
  "type": "integer",
  "unit": "W",
}
`;

exports[`Properties Api one() parse the returned data correctly (with #4) 1`] = `
{
  "categories": [],
  "entities": [
    "material",
  ],
  "id": 4,
  "name": "Conforme",
  "type": "boolean",
}
`;

exports[`Properties Api one() parse the returned data correctly (with #5) 1`] = `
{
  "categories": [],
  "entities": [
    "material",
  ],
  "id": 5,
  "name": "Date d'achat",
  "type": "date",
}
`;

exports[`Properties Api one() parse the returned data correctly (with #6) 1`] = `
{
  "categories": [],
  "entities": [
    "material",
  ],
  "id": 6,
  "is_totalisable": true,
  "name": "Prix d'achat",
  "type": "float",
  "unit": "€",
}
`;

exports[`Properties Api one() parse the returned data correctly (with #7) 1`] = `
{
  "categories": [],
  "entities": [
    "material",
  ],
  "id": 7,
  "name": "Lieu d'usage",
  "options": [
    "Intérieur",
    "Extérieur",
    "Polyvalent",
  ],
  "type": "list",
}
`;

exports[`Properties Api update() parse the returned data correctly (with #1) 1`] = `
{
  "categories": [
    {
      "id": 2,
      "name": "Lumière",
    },
    {
      "id": 1,
      "name": "Son",
    },
  ],
  "entities": [
    "material",
  ],
  "id": 1,
  "is_totalisable": true,
  "name": "Poids",
  "type": "float",
  "unit": "kg",
}
`;

exports[`Properties Api update() parse the returned data correctly (with #2) 1`] = `
{
  "categories": [],
  "entities": [
    "material",
  ],
  "id": 2,
  "max_length": null,
  "name": "Couleur",
  "type": "string",
}
`;

exports[`Properties Api update() parse the returned data correctly (with #3) 1`] = `
{
  "categories": [
    {
      "id": 2,
      "name": "Lumière",
    },
    {
      "id": 1,
      "name": "Son",
    },
  ],
  "entities": [
    "material",
  ],
  "id": 3,
  "is_totalisable": true,
  "name": "Puissance",
  "type": "integer",
  "unit": "W",
}
`;

exports[`Properties Api update() parse the returned data correctly (with #4) 1`] = `
{
  "categories": [],
  "entities": [
    "material",
  ],
  "id": 4,
  "name": "Conforme",
  "type": "boolean",
}
`;

exports[`Properties Api update() parse the returned data correctly (with #5) 1`] = `
{
  "categories": [],
  "entities": [
    "material",
  ],
  "id": 5,
  "name": "Date d'achat",
  "type": "date",
}
`;

exports[`Properties Api update() parse the returned data correctly (with #6) 1`] = `
{
  "categories": [],
  "entities": [
    "material",
  ],
  "id": 6,
  "is_totalisable": true,
  "name": "Prix d'achat",
  "type": "float",
  "unit": "€",
}
`;

exports[`Properties Api update() parse the returned data correctly (with #7) 1`] = `
{
  "categories": [],
  "entities": [
    "material",
  ],
  "id": 7,
  "name": "Lieu d'usage",
  "options": [
    "Intérieur",
    "Extérieur",
    "Polyvalent",
  ],
  "type": "list",
}
`;
